### PR TITLE
Add pre-training transform to convert BatchNorm to BatchNormInternal

### PR DESCRIPTION
**Pull request by pranav-prakash**
**Description**: As opset 14 lacks saved_mean/saved_inv_std, we have to convert nodes to a custom op if we want to ensure good performance on backward pass. This is the second in a series of PRs to implement BN training on CPU (first was #6946).

As part of this we also remove the existing `AdjustBatchNormOutputs` pass – this only works for opset 9 BN and is subsumed by the more general `BatchNormReplacement` we define in this PR.

WIP – I think the only thing missing is a unit test. I'd also appreciate any feedback on whether it's necessary to copy/paste the entire schema-def  for the `BatchNormInternal` op or whether something like protobuf's `toBuilder` exists for these schema defs.

### CODE REVIEW

**Comment by SherlockNoMad** [comment]
Hi @pranav-prakash, thanks for contributing this ! 

Let us know when this is ready for review. I will help you merging this change.

**Comment by pranav-prakash** [comment]
@SherlockNoMad
Thanks, I just added the test so it should be ready for review. One question I had was that the test required me to explicitly import the `com.microsoft` domain in order for it to recognize the `BatchNormInternal` op – I'm not sure if this is expected behavior. When loading a model from disk is `com.microsoft` automatically added as a valid import?

File: orttraining/orttraining/test/optimizer/graph_transform_test.cc [hunk 1]
```diff
@@ -33,6 +34,49 @@ namespace test {
...
+
+  auto rule_transformer_L1 = onnxruntime::make_unique<RuleBasedGraphTransformer>("BatchNormReplacement");
+  rule_transformer_L1->Register(onnxruntime::make_unique<BatchNormReplacement>());
+  onnxruntime::GraphTransformerManager graph_transformation_mgr{5};
+  graph_transformation_mgr.Register(std::move(rule_transformer_L1), TransformerLevel::Level1);
+  ASSERT_STATUS_OK(graph_transformation_mgr.ApplyTransformers(graph, TransformerLevel::Level1, *logger_));
+  
+  ASSERT_TRUE(graph.NumberOfNodes() == 1);
+  // Make sure that BN was updated to add outputs
+  ASSERT_TRUE(graph.Nodes().begin()->MutableOutputDefs().size() == 5);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/optimizer/graph_transform_test.cc:77]: maybe one more assert on the optype to be BacthNormInternal?

File: orttraining/orttraining/core/framework/gradient_graph_builder.cc [hunk 2]
```diff
@@ -32,7 +32,6 @@ GradientGraphBuilder::GradientGraphBuilder(Graph* graph,
   auto rule_based_graph_transformer =
       onnxruntime::make_unique<RuleBasedGraphTransformer>("pre_training_rule_based_graph_transformer");
   rule_based_graph_transformer->Register(make_unique<InsertMaxPoolOutput>());
-  rule_based_graph_transformer->Register(make_unique<AdjustBatchNormOutputs>());
 
   graph_transformation_mgr_.Register(std::move(rule_based_graph_transformer),
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/framework/gradient_graph_builder.cc:36]: InsertMaxPoolOutput can also be removed here. 
actually, graph_transformation_mgr_ can be removed from this class.

**Comment by SherlockNoMad** [comment]
I think we also need to update gradient_builder_registry.cc, to register the gradient builder for BatchNormInternal. 
We should be able to reuse "GetBatchNormalizationGradient" though.

**Comment by SherlockNoMad** [comment]
> @SherlockNoMad
> Thanks, I just added the test so it should be ready for review. One question I had was that the test required me to explicitly import the `com.microsoft` domain in order for it to recognize the `BatchNormInternal` op – I'm not sure if this is expected behavior. When loading a model from disk is `com.microsoft` automatically added as a valid import?

I can imagine msDomain is needed to make the UT work. In the training process, msDomain kernel registry will be automatically included at program start time, coz most of the gradient ops are under MSDomain.

File: orttraining/orttraining/core/optimizer/batchnorm_replacement.cc [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+namespace onnxruntime {
+
+Status BatchNormReplacement::Apply(Graph& graph, Node& bn_node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
+  const auto& bn_inputs = bn_node.MutableInputDefs();
+  auto& bn_outputs = bn_node.MutableOutputDefs();
+  const NodeArg* scale_input_def = bn_inputs[1];
+  auto scale_input_def_type_proto = scale_input_def->TypeAsProto();
+
+  // Guard against a BatchNorm that already has optional outputs present for some reason
+  if (bn_outputs.size() == 1) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/optimizer/batchnorm_replacement.cc:21]: what about the case bn_outputs.size() == 3?

File: orttraining/orttraining/core/optimizer/batchnorm_replacement.cc [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    NodeArg& saved_mean_def = graph.GetOrCreateNodeArg(graph.GenerateNodeArgName("saved_mean_def"), scale_input_def_type_proto);
+    NodeArg& saved_inv_std = graph.GetOrCreateNodeArg(graph.GenerateNodeArgName("saved_inv_std"), scale_input_def_type_proto);
+    bn_outputs.push_back(&saved_inv_std);
+    bn_outputs.push_back(&saved_mean_def);
+  }
+
+  // check Batch Normalization node has 5 output node args for training mode
+  ORT_ENFORCE(bn_node.OutputDefs().size() == 5);
+
+  Node& batchnorm_internal_node = graph.AddNode(graph.GenerateNodeName("BatchNormInternal"),
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/optimizer/batchnorm_replacement.cc:36]: would be good to preserve the original node name as a substring of the new name
so that we can track the original of node

File: orttraining/orttraining/test/optimizer/graph_transform_test.cc [hunk 5]
```diff
@@ -33,6 +34,49 @@ namespace test {
...
+  scale_tensor_type.mutable_tensor_type()->set_elem_type(TensorProto_DataType_FLOAT);
+  scale_tensor_type.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value(3);
+
+  auto& input_X = graph.GetOrCreateNodeArg("X", &input_tensor_type);
+  auto& input_scale = graph.GetOrCreateNodeArg("scale", &scale_tensor_type);
+  auto& input_B = graph.GetOrCreateNodeArg("B", &scale_tensor_type);
+  auto& input_mean = graph.GetOrCreateNodeArg("input_mean", &scale_tensor_type);
+  auto& input_var = graph.GetOrCreateNodeArg("input_var", &scale_tensor_type);
+
+  auto& output_Y = graph.GetOrCreateNodeArg("Y", &input_tensor_type);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/optimizer/graph_transform_test.cc:63]: maybe another test case with 3 outputs?

**Comment by SherlockNoMad** [comment]
Also need to add BatchNormInternal into the STOP_GRADIENT_EDGES in gradient_graph_builder.h

**Comment by SherlockNoMad** [comment]
Thanks again for the PR. The bulk part is looking good! 
We should be good to merge once the comments are addressed

File: orttraining/orttraining/test/optimizer/graph_transform_test.cc (hunk 1 above)
Comment (pranav-prakash) [reply orttraining/orttraining/test/optimizer/graph_transform_test.cc:77]: Done

File: orttraining/orttraining/core/framework/gradient_graph_builder.cc (hunk 2 above)
Comment (pranav-prakash) [reply orttraining/orttraining/core/framework/gradient_graph_builder.cc]: Done.

File: orttraining/orttraining/core/optimizer/batchnorm_replacement.cc (hunk 3 above)
Comment (pranav-prakash) [reply orttraining/orttraining/core/optimizer/batchnorm_replacement.cc:20]: Fixed, thanks for catching this!

File: orttraining/orttraining/core/optimizer/batchnorm_replacement.cc (hunk 4 above)
Comment (pranav-prakash) [reply orttraining/orttraining/core/optimizer/batchnorm_replacement.cc]: Fixed, I'm doing `bn_node.Name() + "_BatchNormInternal"` which should hopefully suffice

File: orttraining/orttraining/test/optimizer/graph_transform_test.cc (hunk 5 above)
Comment (pranav-prakash) [reply orttraining/orttraining/test/optimizer/graph_transform_test.cc:63]: Done. Added a testcase for opset14 with 3 outputs, as well as opset 9 with 5 outputs

**Comment by pranav-prakash** [comment]
Thanks for the quick review! I addressed the comments you left.

>register gradient builder for BatchNormInternal.
Done. I'll add a CPU implementation of the gradient in a subsequent PR.

I'll also register an implementation for BatchNormInternal (re-using existing BatchNorm kernel) in a subsequent PR. I believe there's also a pre-existing CUDA kernel for this, but it seems to be broken (tests were disabled) so someone else who's both more familiar and capable of testing it will have to modify those parts.

**Review by SherlockNoMad — APPROVED** [review]

**Comment by SherlockNoMad** [comment]
sure. Appreciate it if you can contribute the CPU BatchNormGradient kernel !!

**Comment by pranav-prakash** [comment]
@SherlockNoMad
Seems like maybe we can't remove `InsertMaxPoolOutput ` in `gradient_graph_builder.cc` after all? It fails the test
```
1: C++ exception with description "/onnxruntime_src/orttraining/orttraining/core/graph/gradient_builder_base.h:102 onnxruntime::training::ArgDef onnxruntime::training::GradientBuilderBase::O(size_t) const i < node_->OutputDefs().size() was false. 
```

although I'm not sure why this is since we register it in `graph_transformer_utils.cc` as well.

**Comment by SherlockNoMad** [comment]
ah... ic.. the GradientCheckerTest.MaxPoolGrad is not invoking the GeneratePreTrainingTransformers, as it's taking the InferenceSession path to construct the test model. 

Please revert the Maxpool related changes, we will fix this in the latter PRs. Sorry for the inconvenience caused.

**Comment by pranav-prakash** [comment]
@SherlockNoMad Done, thanks for determining the cause!
